### PR TITLE
Chameleon Eyewear

### PR DESCRIPTION
**Pull request by ghost**
Syndicate thermals can now be toggled to look like any kind of eyewear you want. This will hopefully help avoid the "why the hell does that med doctor have mesons" type situations.

Added chameleon security HUDs, which share the same chameleon function.

Gang leaders and revheads now spawn with the chameleon sec HUDs so they aren't left standing awkwardly with pen/flash in hand after trying to convert an assistant with a loyalty implant.

The chameleon proc looks kind of ugly right now, I'm not sure how to remove all the bad/generic eyewear (like shadowling eyes) from the list of choices and make it look pretty. Help with that would be appreciated.

### CODE REVIEW

File: code/modules/clothing/glasses/glasses.dm [hunk 1]
```diff
@@ -230,3 +232,27 @@
...
 	item_state = "redglasses"
 
+
+/obj/item/clothing/glasses/proc/chameleon(var/mob/user)
+	var/list/glasses_types = typesof(/obj/item/clothing/glasses) - src.type - /obj/item/clothing/glasses/hud - /obj/item/clothing/glasses/night/shadowling - /obj/item/clothing/glasses/hud/toggle - /obj/item/clothing/glasses
+	var/list/glasses = list()
+
+	// Generate them into a list
+	for(var/glasses_type in glasses_types)
+		var/obj/item/clothing/glasses/S = new glasses_type
```
Comment (RemieRichards) [inline code/modules/clothing/glasses/glasses.dm]: please no.

index the TYPE by the name:
```
for(var/obj/item/clothing/glasses/G in glasses_types)
    glasses[initial(G.name)] = G
```
then see below

File: code/modules/clothing/glasses/glasses.dm [hunk 2]
```diff
@@ -230,3 +232,27 @@
...
+		glasses[capitalize(S.name)] = S
+
+	var/list/show_glasses = list("EXIT" = null) + sortList(glasses) // the list that will be shown to the user to pick from
+
+	var/input_glasses = input(user, "Choose a piece of eyewear to disguise as.", "Choose glasses style.") in show_glasses
+
+	if(user && src in user.contents)
+
+		var/obj/item/clothing/glasses/chosen_glasses = glasses[capitalize(input_glasses)]
```
Comment (RemieRichards) [inline code/modules/clothing/glasses/glasses.dm]:
```
if(glasses[capitalize(input_glasses)])
    var/obj/item/clothing/glasses/chosen_glasses = new glasses[capitalize(input_glasses)]()
```

File: code/modules/clothing/glasses/glasses.dm [hunk 3]
```diff
@@ -230,3 +232,27 @@
 	icon_state = "redglasses"
 	item_state = "redglasses"
 
+
+/obj/item/clothing/glasses/proc/chameleon(var/mob/user)
+	var/list/glasses_types = typesof(/obj/item/clothing/glasses) - src.type - /obj/item/clothing/glasses/hud - /obj/item/clothing/glasses/night/shadowling - /obj/item/clothing/glasses/hud/toggle - /obj/item/clothing/glasses
```
Comment (RemieRichards) [inline code/modules/clothing/glasses/glasses.dm]: Some of these (not all) can be removed from this line, and solved in the loop below, by checking for 
```if(initial(G.flags) & ABSTRACT```

**Comment by SconesC** [comment]
>Gang leaders and revheads now spawn with the chameleon sec HUDs so they aren't left standing awkwardly with pen/flash in hand after trying to convert an assistant with a loyalty implant.

Thanks but no thanks, these gamemodes are already balanced in this regard. Knowing who has implants, who does not, and if you're set to wanted are all HUGE advantages to a revhead (and a gang boss to a somewhat lesser extent)

The Cham Thermals are cool. Giving Cham Sechuds to every revhead and gang boss, not so cool.

File: code/game/gamemodes/revolution/revolution.dm [hunk 4]
```diff
@@ -137,10 +138,16 @@
 		"right hand" = slot_r_hand,
 	)
 	var/where = mob.equip_in_one_of_slots(T, slots)
+	var/where2 = mob.equip_in_one_of_slots(C, slots)
 	mob.equip_in_one_of_slots(R,slots)
 
 	mob.update_icons()
 
+	if (!where2)
+		mob << "The Syndicate were unfortunately unable to get you chameleon security HUD."
```
Comment (RemieRichards) [inline code/game/gamemodes/revolution/revolution.dm]: "get you <B>A</B> chameleon security HUD."

**Comment by WJohn** [comment]
I think it'd be nice if you could just use the thermals in your hand to repick what they look like. So you'd be able to make em look like sunglasses, secglasses, medhuds, mesons etc.

File: code/modules/clothing/glasses/glasses.dm (hunk 2 above)
Comment (KorPhaeron) [inline code/modules/clothing/glasses/glasses.dm]: Says it's an invalid expression. I assume I'm doing something wrong but double checking with you to make sure you wrote it correctly.

**Comment by KorPhaeron** [comment]
@WJohn 

Exactly how it works.

File: code/modules/clothing/glasses/glasses.dm (hunk 2 above)
Comment (RemieRichards) [inline code/modules/clothing/glasses/glasses.dm]: it requires you wrote what I did in the comment above this one.

**Comment by WJohn** [comment]
Are you able to repick it over and over or is it once like agent ID cards were for the longest time?

**Comment by KorPhaeron** [comment]
@WJohn

Repick forever

**Comment by PKPenguin321** [comment]
awesome :+1: 

is there something like this for noslips yet?

**Comment by KorPhaeron** [comment]
I gave up and redid it with a switch/manually set a series of glasses to choose from cause I couldn't get the old way to work, let me know if there are any glasses that are missing that you'd like added to the list.

**Comment by KorPhaeron** [comment]
@SconesC 

In cases where security is losing, it won't matter anyway.

In cases where security catches them early, it will give revs a chance to bounce back and nab crewmembers that sec hasn't kidnapped for implanting yet. Right now security can see anyone who is implanted/not implanted, while revs can't do the same. Security shouldn't have a straight up advantage in terms of conversion vs revs when it's their entire gimmick.

**Comment by Durkel** [comment]
I like the idea especially with the thermals, but I don't think giving revs/gangs a hud off the bat is a good idea. Just my two cents.

**Comment by Lumipharon** [comment]
Just make it purchasable for gang leaders.

**Comment by KorPhaeron** [comment]
@Lumipharon 

1:23 PM - Ikarrus: I think making them spawn with them would be cool
1:23 PM - Ikarrus: I kind of want to shy away from giving them too many tools to buy, because then it just becomes like uplinks with most of the items ignored
1:23 PM - Kor: fair enough
1:24 PM - Kor: Everyone buying them would just make anyone with implants get hunted and murdered immediatley anyway
1:24 PM - Ikarrus: this way only the bosses/revheads would see
1:24 PM - Kor: nobody else in the gang really needs to know
1:24 PM - Ikarrus: I like that. Maybe they'll work together more often
1:24 PM - Ikarrus: probably not
1:24 PM - Ikarrus: but I can wish

Disclaimer: I think Ikarrus was under the impression they'd view implants only when I discussed this change with him, but I got lazy and didn't want to make new HUD code

**Comment by Supermichael777** [comment]
sec huds are a powerful thing especially after you convert that all access clown/hop and can suddenly used the bots ability to hand out stuns and distract cops with radio spam.

**Comment by KorPhaeron** [comment]
I looked further at the HUD code, would actually be fairly simple to make the revhead/gangleader thing only show implants if that would make people happier.

**Comment by Firecage** [comment]
Question though. Can't you simply make them use the items name/icon/desc/etc which it changes to instead of manually copy'ing/typing it over. Since the latter seems redundant.

**Comment by Shadowlight213** [comment]
:+1: to more chameleon items. The chameleon jumpsuit is kinda worthless currently since we have things like departmental specific boots, backpacks, etc..

**Comment by AlexanderUlanH** [comment]
Chameleon voice changer when?

**Comment by SconesC** [comment]
@KorPhaeron If you really want to expand on the chameleon gear, a chameleon kit (eyegear, shoes, uniform) would really be the best considering shoes and shit are the easiest way to bust a disguised traitor on the go

**Comment by KorPhaeron** [comment]
18 days now, could this be accepted or denied @RemieRichards @phil235  @Cheridan @Miauw62

**Comment by SconesC** [comment]
Gang balance has changed a lot and they didn't need this at the time of PR creation nor do they need it now

**Comment by Razharas** [comment]
Eh i like it

**Comment by SconesC** [comment]
Eughhhhhhhhhhh

**Pull request merged**